File: Leetcode/dfs_recursion_backtracking/10_RegularExpressionMatching/Solution.py
class Solution(object):
    def isMatch(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: bool
        """
        if not p:
            return not s
        if not s:
            return (not p) or self.isMatch(s, p[2:]) if len(p)>1 and p[1]=='*' else False
        if p[0]!='.' and s[0]!=p[0]:
            return self.isMatch(s, p[2:]) if len(p)>1 and p[1]=='*' else False
        if len(p)>1 and p[1]=='*':
            return self.isMatch(s[1:], p) or self.isMatch(s, p[2:])
        return self.isMatch(s[1:], p[1:])

File: Leetcode/dfs_recursion_backtracking/10_RegularExpressionMatching/Solution.java
class Solution {
    public boolean isMatch(String s, String p) {
        if(p.length()==0)
            return s.length()==0;
        if(s.length()==0)
            return (p.length()==0)||
            (p.length()>1 && p.charAt(1)=='*')?isMatch(s, p.substring(2)): false;
        if( p.charAt(0)!='.' && s.charAt(0)!=p.charAt(0) )
            return (p.length()>1 && p.charAt(1)=='*')? isMatch(s, p.substring(2)): false;
        if( p.length()>1 && p.charAt(1)=='*')
            return isMatch(s.substring(1), p)||isMatch(s, p.substring(2));
        return isMatch(s.substring(1), p.substring(1));
    }
}
